pkg/controller/compute/v1beta1/cluster.go

```diff
@@ -0,0 +1,287 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	runtimev1alpha1 "github.com/crossplaneio/crossplane-runtime/apis/core/v1alpha1"
+	"github.com/crossplaneio/crossplane-runtime/pkg/logging"
+	"github.com/crossplaneio/crossplane-runtime/pkg/meta"
+	"github.com/crossplaneio/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplaneio/stack-gcp/apis/compute/v1beta1"
+	gcpv1alpha3 "github.com/crossplaneio/stack-gcp/apis/v1alpha3"
+	gcp "github.com/crossplaneio/stack-gcp/pkg/clients"
+	gke "github.com/crossplaneio/stack-gcp/pkg/clients/container"
+)
+
+const (
+	controllerName    = "gke.compute.gcp.crossplane.io"
+	finalizer         = "finalizer." + controllerName
+	clusterNamePrefix = "gke-"
+
+	requeueOnWait   = 30 * time.Second
+	requeueOnSucces = 2 * time.Minute
+
+	updateErrorMessageFormat         = "failed to update cluster object: %s"
+	erroredClusterErrorMessageFormat = "gke cluster is in %s state with message: %s"
+)
+
+// Error strings.
+const (
+	errGetProvider             = "cannot get Provider"
+	errGetProviderSecret       = "cannot get Provider Secret"
+	errNewClient               = "cannot create new GKE cluster client"
+	errManagedUpdateFailed     = "cannot update GKECluster custom resource"
+	errNotCluster              = "managed resource is not a GKECluster"
+	errUpdateCR                = "cannot update GKECluster custom resource"
+	errGetCluster              = "cannot get GKE cluster"
+	errCreateCluster           = "cannot create GKE cluster"
+	errUpdateCluster           = "cannot update GKE cluster"
+	errDeleteCluster           = "cannot delete GKE cluster"
+	errDeleteBootstrapNodePool = "cannot delete bootstrap node pool"
+)
+
+// Amounts of time we wait before requeuing a reconcile.
+const (
+	aLongWait         = 60 * time.Second
+	parentFormat      = "projects/%s/locations/%s"
+	clusterNameFormat = "projects/%s/locations/%s/clusters/%s"
+)
+
+// Error strings
+const (
+	errUpdateManagedStatus = "cannot update managed resource status"
+)
+
+var (
+	log           = logging.Logger.WithName("controller." + controllerName)
+	ctx           = context.Background()
+	result        = reconcile.Result{}
+	resultRequeue = reconcile.Result{Requeue: true}
+)
+
+// GKEClusterController is responsible for adding the GKECluster
+// controller and its corresponding reconciler to the manager with any runtime configuration.
+type GKEClusterController struct{}
+
+// SetupWithManager creates a new Controller and adds it to the Manager with default RBAC. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func (c *GKEClusterController) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1beta1.GKEClusterGroupVersionKind),
+		resource.WithExternalConnecter(&clusterConnector{kube: mgr.GetClient(), newServiceFn: container.NewService}),
+		resource.WithManagedInitializers())
```

Comment:
Overriding default `NewManagedNameAsExternalName` initializer.

Reply:
Do we need to do this?

I notice we're using the fully qualified name of the cluster (e.g. `projects/coolproj/locations/us-central1/clusters/mycluster`) rather than the unqualified name (e.g. `mycluster`). Are you perhaps thinking of the external name as meaning "the value that must be passed to the external client to identify the thing"? It's more intended to be "a setting the user can specify to control the name of their cluster as it appears in the GCP console".

My recommendation would be to use the unqualified name as the external name, use the regular external name initializer, and pass the result of some function with a signature like like `GetFullyQualifiedName(mg)` to the GKE client.

Reply:
I think I originally implemented this prior to adding the `Location` field, but you are definitely correct in that we can and should use the regular external name initializer here. Thanks for pointing that out!